### PR TITLE
[None][fix] Survive a failed forward step in the executor loops

### PR DESCRIPTION
**Pull request by brnguyen2**
## Dev Engineer Review

- Adds multimodal admission validation through `validate_multimodal_request_data`.
- Updates executor loops to handle non-fatal `_forward_step` failures.
- Raises `RuntimeError` on the last pipeline-parallel rank when forward output is missing.
- Adds warnings when decode concurrency exceeds CUDA graph coverage.
- Review finding counts are unavailable.

## QA Engineer Review

- Adds CPU-only tests for executor failure handling, pipeline-parallel errors, multimodal validation, and CUDA graph warnings.
- Updates chunked-transfer request initialization.
- Coverage verdict: **needs follow-up** because test execution results are not supplied.
- No changed integration test-list or manual-QA list files were identified.

## Per-File QA Perspective

- `tensorrt_llm/_torch/models/modeling_multimodal_mixin.py`: Adds request admission validation. Verify accepted and rejected multimodal requests.
- `tensorrt_llm/_torch/pyexecutor/_util.py`: Adds CUDA graph coverage warnings. Verify thresholds and suppression when graphs are disabled.
- `tensorrt_llm/_torch/pyexecutor/py_executor.py`: Changes forward-failure handling and multimodal validation. Verify collectives, error paths, and pipeline-parallel behavior.
- `tests/unittest/_torch/executor/kv_cache/test_kv_cache_budget_split.py`: Tests CUDA graph coverage warning cases. Test-list registration was not verified.
- `tests/unittest/executor/test_forward_error_survival.py`: Tests forward-failure survival, pipeline-parallel errors, and multimodal validation. Test-list registration was not verified.
- `tests/unittest/disaggregated/test_chunked_transfer.py`: Tests request validation after multimodal data initialization. Test-list registration was not verified.

## Description

A non-fatal forward error is supposed to be survivable: `_forward_step` catches the exception, fails the affected requests via `_handle_errors` (which enqueues their error responses) and returns `None`. But none of the executor loops checked for that `None`, so the next statement dereferenced `batch_outputs['logits']`, crashed the event loop on every rank, and the whole engine died behind a correct-looking client error response.

This PR has two commits:

**Commit 1 — survive a failed forward step:**
- The overlap and non-overlap loops handle a `None` forward result by skipping the iteration's sampling and bookkeeping (the failed requests already got their error responses) while keeping the rank-symmetric flush collectives in place.
- The PP loop raises an explicit `RuntimeError` on the last PP rank instead of the opaque `TypeError`: inter-PP peers are inside inter-PP communication for the microbatch and cannot skip it in lockstep, so the iteration is not survivable there.
- A new `MultimodalModelMixin.validate_multimodal_request_data` hook, called from `PyExecutor._validate_request`, lets a model reject a bad multimodal payload (e.g. an image whose vision-attention segment exceeds the encoder budget) at admission time. The same condition raised inside the model forward is handled batch-wide and fails every request scheduled alongside the bad one; rejecting at admission fails only that request.

**Commit 2 — warn when admitted decode concurrency outruns CUDA graph coverage:**
The configured decode concurrency (scheduler `max_batch_size`, and any KV-capacity ceiling a cache manager reports via a worst-case per-request cost) can sit well past the range covered by captured CUDA graphs. Decode batches above that range run in eager mode, whose per-iteration latency is far higher than a graph replay, so the extra concurrency can cost throughput instead of buying it (measured: admitted ~86 vs graphs to 32 took the mean decode iteration from 65 ms to 1610 ms, ~5x output-throughput collapse). Coverage is not the largest captured size: without `cuda_graph_config.enable_padding` a decode batch replays a graph only at an exactly captured size, so the covered range is the dense `{1..n}` prefix of the capture set. Emit a one-time engine-init warning naming the remedies (raise `cuda_graph_config.max_batch_size` / `batch_sizes` with `enable_padding`, or lower `max_batch_size`). A warning rather than a refusal, because all-eager decode is a legitimate choice at low offered load; silent when CUDA graphs are disabled outright.

## Test Coverage

- `tests/unittest/executor/test_forward_error_survival.py` (CPU-only, new): the admission hook contract (default accept, override reject), the `_validate_request` call site (reject via hook, text-only skip, hook-less models), and loop-level `None` handling — the non-overlap and overlap loops complete the iteration and keep running without sampling, and the PP loop raises the explicit fatal — via stubbed executor shells driving the real loop bodies.
- `tests/unittest/_torch/executor/kv_cache/test_kv_cache_budget_split.py` (CPU-only): `TestGraphCaptureConcurrencyWarning` covers the warning trigger, the no-padding dense-prefix coverage semantics, padded coverage suppressing the warning, scheduler-capped admission, and graphs-disabled silence.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

### CODE REVIEW

File: tests/unittest/_torch/executor/kv_cache/test_kv_cache_budget_split.py [hunk 1]
```diff
@@ -1175,3 +1175,94 @@ def test_non_external_drafter_is_untouched(self, mocker):
         assert c._speculative_config.spec_dec_mode == SpeculativeDecodingMode.MTP_EAGLE_ONE_MODEL
 
         assert c._get_draft_kv_model_config() is draft_model_config
+
+
+class TestGraphCaptureConcurrencyWarning:
```
[the remark this answers is not in the document]
Comment (brnguyen2) [reply tests/unittest/_torch/executor/kv_cache/test_kv_cache_budget_split.py:1259]: Correct, and the cause is upstream of this PR. tests/integration/test_lists/waives.txt has a directory-level `unittest/_torch/executor SKIP` entry left behind by #19381 after #19382 fixed the bug it cites, so every test in that tree has been skipped in CI since 2026-09-18, not only this class. #19533 replaces that line with per-test waives for the tests that actually fail on main, so the rest of the tree (including this class) runs again in the CPU unit-test stages.

I am keeping the test where it is: it belongs with the executor tests, and moving it to dodge a stale waive would leave the tree hidden. Verified at this head on a B200 box with the blanket waive removed: TestGraphCaptureConcurrencyWarning 6 passed, 0 failed. Once #19533 merges and this PR rebases, the class runs in CI.

File: tensorrt_llm/_torch/pyexecutor/_util.py [hunk 2]
```diff
@@ -1912,6 +1912,94 @@ def _compute_draft_budget_shares(
...
+        if isinstance(self._max_batch_size, int) and self._max_batch_size > 0:
+            ceilings.append(self._max_batch_size)
+        for budget, cost in ((target_budget, target_kv), (draft_budget,
+                                                          draft_kv)):
+            # ``bytes_per_request`` is an optional CacheCost extension (a
+            # manager-reported worst-case per-request pool spend); absent, the
+            # scheduler max_batch_size is the only admission ceiling.
+            per_request = getattr(cost, "bytes_per_request", None)
+            if isinstance(per_request, int) and per_request > 0:
+                ceilings.append(max(0, budget - cost.intercept) // per_request)
```
[the remark this answers is not in the document]
Comment (brnguyen2) [reply tensorrt_llm/_torch/pyexecutor/_util.py:2161]: Reasonable case to cover, but I would rather keep this PR to the survival fix. The class cannot run in CI at all until #19533 lands (see the thread on the test file), and the KV-capacity ceiling needs the manager cost path stubbed, which the existing helper does not do. I will add the bytes_per_request ceiling case in the follow-up that updates the budget_split fixtures.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 3]
```diff
@@ -5198,7 +5222,31 @@ def _executor_loop_overlap(self):
...
+                        # so fall back to the empty-batch bookkeeping for the
+                        # rest of this iteration instead of dereferencing
+                        # batch_outputs['logits'] and taking down every rank.
+                        # The previous batch's requests were failed with the
+                        # rest of active_requests, so drop it rather than
+                        # updating terminated requests.
+                        can_queue = False
+                        can_queue_this_rank = False
+                        should_process_previous_batch = False
+                        self.previous_batch = None
```
[the remark this answers is not in the document]
Comment (brnguyen2) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py:5247]: Good catch. Dropping `previous_batch` there skips `_process_iter_stats`, which is what normally returns the borrowed pair, so those events were leaking. Fixed in 1a582b7a52: the branch now releases the previous batch's pair before clearing it, and `test_executor_loop_overlap_survives_a_failed_forward` seeds a previous batch with pool-borrowed events and asserts the release. The non-overlap loop already released its pair on the None path (`sample_state` stays None so the `elif` release fires), and the PP loop raises, so that was the only spot.

File: tensorrt_llm/_torch/pyexecutor/_util.py [hunk 4]
```diff
@@ -2092,6 +2092,94 @@ def _compute_draft_budget_shares(
...
+        if not ceilings:
+            return
+        admissible = min(ceilings)
+        if admissible <= self._CUDA_GRAPH_CAPTURE_WARN_FACTOR * covered:
```
[the remark this answers is not in the document]
Comment (brnguyen2) [reply tensorrt_llm/_torch/pyexecutor/_util.py:2165]: Added both sides of the boundary in 1a582b7a52: with dense coverage up to 32 and factor 1.5, `max_batch_size=48` produces no warning and `49` produces exactly one (asserting `~49` and `up to 32` in the message). Same `_creator()`/`_warnings()` helpers as the rest of the class.
